plugins/lua/dwarfmonitor.lua

```diff
@@ -146,6 +146,95 @@ function Widget_cursor:render_body(p)
     p:string(self.output)
 end
 
+-- FPS counter that doesn't count frames when paused,
+-- so FPS doesn't have to slowly stabilize after unpausing.
+Widget_fps = defclass(Widget_fps, Widget)
+Widget_fps.ATTRS = {
+    output = '',
+    show_fps_in_init = df.global.gps.display_frames == 1,
+    prev_paused = false,
+    prev_clock = nil,
+    prev_frames = nil,
+    elapsed_clock = 0.0,
+    elapsed_frames = 0,
+    history = {},
+}
+function Widget_fps:fps()
+    local tps = self.elapsed_frames / self.elapsed_clock
+    local count = 1
+
+    if not (tps >= 0.0) then -- ignore NaNs
+        tps = 0
+        count = 0
+    end
+
+    for i = 1, self.opts.average_over_seconds do
+        if self.history[i] and self.history[i] >= 0.0 then -- ignore NaNs
+            tps = tps + self.history[i]
+            count = count + 1
+        end
+    end
+    return tps / count
+end
+function Widget_fps:updateOutput()
+    local fps = self:fps()
+    if not (fps >= 0.0) then -- NaN
+        self.output = "   -"
+    elseif fps <= 1.0 then -- less than 1 fps
+        self.output = string.format("%0.2f", fps)
+    else
+        self.output = string.format("%4d", math.floor(0.5 + fps))
+    end
+    self.output = "FPS:" .. self.output
+end
+function Widget_fps:update()
+    -- dont show fps counter unless [FPS:YES] is set in init.txt
+    if not self.show_fps_in_init then
+        return
+    end
+
+    -- hide internal FPS counter because it is drawn on top of this widget
+    df.global.gps.display_frames = 0
+
+    if not self.opts.average_over_seconds then
+        self.opts.average_over_seconds = 5
+    end
+
+    if df.global.pause_state then
+        self.prev_paused = true
+    else
+        local clock = os.clock()
```

Comment:
This is not portable - it returns the elapsed CPU time on Linux and macOS, not the absolute time, so I'm seeing stuff like "FPS: 342" and "FPS: 887" with a limit of 100.